### PR TITLE
Measure latency of network messages assuming all nodes have synchronized clocks

### PR DESCRIPTION
**Pull request by nikurt**
This PR measures latency of all messages by adding a timestamp at the lowest level - at the protobuf level.

cc: @mm-near @mzhangmzz 
CPR-86

### CODE REVIEW

**Comment by nagisa** [comment]
The overall approach makes sense to me, I had done something similar in the past. I’ve a question though: we have some beginnings of a distributed tracing framework, is there a reason why we couldn’t or shouldn’t utilize that for something like this?

**Comment by nagisa** [comment]
> distributed tracing framework will let us see the network latency of a single trace, but doesn't support extracting network latency across all available traces and displaying them as a timeseries. In Grafana Tempo we can see timeseries of duration of traces filtered by span_type, etc. But network latency isn't represented as a span, but rather as a gap between spans.

Right, I wonder if we could somehow encode this as a key-value in the receiver’s span rather than as a gap between spans. I believe it should be possible to graph the value of a specific KV. But given the 2nd point, it seems like metrics are the way to go here anyway.

File: chain/network/src/network_protocol/proto_conv.rs [hunk 1]
```diff
@@ -397,6 +400,7 @@ impl TryFrom<&proto::Block> for Block {
 
 impl From<&PeerMessage> for proto::PeerMessage {
     fn from(x: &PeerMessage) -> Self {
+        let tx_timestamp = SystemTime::now().duration_since(UNIX_EPOCH);
```
Comment (pompon0) [inline chain/network/src/network_protocol/proto_conv.rs:403]: please don't make time-measurement encoding-dependent and move it to peer.rs.
Here you should just accept a timestamp as an argument.

File: chain/network/src/network_protocol/network.proto [hunk 2]
```diff
@@ -343,4 +343,6 @@ message PeerMessage {
     EpochSyncFinalizationRequest epoch_sync_finalization_request = 22;
     EpochSyncFinalizationResponse epoch_sync_finalization_response = 23;
   }
+
+  google.protobuf.Timestamp tx_timestamp = 25;
```
Comment (pompon0) [inline chain/network/src/network_protocol/network.proto:347]: Please document what this field represents.

File: chain/network/src/network_protocol/proto_conv.rs [hunk 3]
```diff
@@ -483,6 +487,13 @@ impl From<&PeerMessage> for proto::PeerMessage {
                     })
                 }
             }),
+            tx_timestamp: MF::from_option(tx_timestamp.ok().map(|duration| {
```
Comment (pompon0) [inline chain/network/src/network_protocol/proto_conv.rs:490]: please use https://github.com/near/nearcore/blob/2eec03616a924112da587a4da676c4883d1acd27/chain/network-primitives/src/time.rs#L66 
and convert it like this: https://github.com/near/nearcore/pull/7023/files#diff-2e879b433bc841c07dc50b3c7947ebf60988834054e0dd39b74e173b12b3c99d
(unfortunately this PR is not merged yet).

File: chain/network/src/network_protocol/proto_conv.rs [hunk 4]
```diff
@@ -605,6 +616,21 @@ impl TryFrom<&proto::PeerMessage> for PeerMessage {
...
                 ))
             }
-        })
+        };
+
+        if let Some(tx_timestamp) = x.tx_timestamp.as_ref() {
+            let tx_timestamp =
+                UNIX_EPOCH.add(Duration::new(tx_timestamp.seconds, tx_timestamp.nanos));
+            let rx_timestamp = SystemTime::now();
+            let latency = rx_timestamp.duration_since(tx_timestamp);
```
Comment (pompon0) [inline chain/network/src/network_protocol/proto_conv.rs:625]: ditto, move it to peer_actor.rs

**Review by pompon0 — COMMENTED** [review]
In the context of the jira issue, aren't we more interested in latency of routed messages, rather than p2p messages?

File: chain/network/src/network_protocol/network.proto (hunk 2 above)
Comment (pompon0) [inline chain/network/src/network_protocol/network.proto:347]: "_timestamp" suffix is redundant with the field type. How about sth like "sent_at"?.

**Comment by nikurt** [comment]
Please take another look. Moved the timestamp to the business logic layer and the change doesn't need a protocol upgrade.

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 5]
```diff
@@ -2275,6 +2281,19 @@ impl PeerManagerActor {
         }
     }
 
+    // The routed message received its destination. If the timestamp of creation of this message is
+    // known, then update the corresponding latency metric histogram.
+    fn record_routed_msg_latency(&self, msg: &RoutedMessageV2) {
+        if let Some(created_at) = &msg.created_at {
+            if let Ok(created_at) = time::UtcSerializable::to_instant(created_at) {
+                let duration = self.clock.now_utc().sub(created_at);
```
Comment (pompon0) [inline chain/network/src/peer_manager/peer_manager_actor.rs:2289]: yes, that's the way I intended the clock to be used :+1: . Please also use self.clock when constructing the RoutedMessageV2.

**Review by pompon0 — CHANGES_REQUESTED** [review]
LGTM, I've left just a few comments to dedup logic that I also implemented in another PR, and to use the clock more consistently. Thanks.

File: chain/network/src/network_protocol/proto_conv/peer_message.rs [hunk 6]
```diff
@@ -265,9 +266,18 @@ impl TryFrom<&proto::PeerMessage> for PeerMessage {
...
-            ProtoMT::Routed(r) => PeerMessage::Routed(Box::new(
-                RoutedMessage::try_from_slice(&r.borsh).map_err(Self::Error::Routed)?,
-            )),
+            ProtoMT::Routed(r) => PeerMessage::Routed(Box::new(RoutedMessageV2 {
+                msg: RoutedMessage::try_from_slice(&r.borsh).map_err(Self::Error::Routed)?,
+                created_at: r.created_at.as_ref().and_then(|timestamp| {
+                    match utc_from_proto(timestamp) {
+                        Ok(utc) => Some(utc),
+                        Err(err) => {
+                            tracing::warn!(target: "network", err=?err, "Malformed routed message created_at timestamp");
```
Comment (pompon0) [inline chain/network/src/network_protocol/proto_conv/peer_message.rs:275]: nit: I'd return an error if timestamp is not parseable: it is allowed to be missing, but not incorrect.

**Review by pompon0 — APPROVED** [review]
LG!

**Review by nagisa — APPROVED** [review]

**Pull request merged**